src/Cli/dotnet/ToolPackage/ToolPackageDownloader.cs

```diff
@@ -0,0 +1,244 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using Microsoft.DotNet.Cli.NuGetPackageDownloader;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.ToolPackage;
+using Microsoft.DotNet.Tools;
+using Microsoft.Extensions.EnvironmentAbstractions;
+using NuGet.Client;
+using NuGet.Common;
+using NuGet.ContentModel;
+using NuGet.Frameworks;
+using NuGet.LibraryModel;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.ProjectModel;
+using NuGet.Repositories;
+using NuGet.RuntimeModel;
+using NuGet.Versioning;
+
+
+namespace Microsoft.DotNet.Cli.ToolPackage
+{
+    internal class ToolPackageDownloader:IToolPackageDownloader
+    {
+        private INuGetPackageDownloader _nugetPackageDownloader;
+        private readonly IToolPackageStore _toolPackageStore;
+
+        protected DirectoryPath _toolDownloadDir;
+        protected DirectoryPath _toolReturnPackageDirectory;
+        protected DirectoryPath _toolReturnJsonParentDirectory;
+
+        protected readonly DirectoryPath _globalToolStageDir;
+        protected readonly DirectoryPath _localToolDownloadDir;
+        protected readonly DirectoryPath _localToolAssetDir;
+        
+
+        public ToolPackageDownloader(
+            IToolPackageStore store
+        )
+        {
+            _toolPackageStore = store ?? throw new ArgumentNullException(nameof(store)); ;
+            _globalToolStageDir = _toolPackageStore.GetRandomStagingDirectory();
+            _localToolDownloadDir = new DirectoryPath(Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "nuget", "package"));
+            _localToolAssetDir = new DirectoryPath(PathUtilities.CreateTempSubdirectory());  
+        }
+
+        public IToolPackage InstallPackageAsync(PackageLocation packageLocation, PackageId packageId,
+            VersionRange versionRange = null,
+            string targetFramework = null,
+            string verbosity = null,
+            bool isGlobalTool = false
+            )
+        {
+            _toolDownloadDir = isGlobalTool ? _globalToolStageDir : _localToolDownloadDir;
+            var assetFileDirectory = isGlobalTool ? _globalToolStageDir : _localToolAssetDir;
+            _nugetPackageDownloader = new NuGetPackageDownloader.NuGetPackageDownloader(_toolDownloadDir);
+
+            NuGetVersion version = DownloadAndExtractPackage(packageLocation, packageId, _nugetPackageDownloader, _toolDownloadDir.Value).GetAwaiter().GetResult();
+            CreateAssetFiles(packageId, version, _toolDownloadDir, assetFileDirectory);
```

Comment:
Is the assets file actually used?  If all you need is the `DotnetToolSettings.xml` file, you could reduce a lot of complexity, and improve runtime performance, by searching `Directory.EnumerateFiles`, using `DotnetToolSettings.xml` as the search pattern, then using a regex to figure out all TFMs and RIDs, then finally use NuGet's APIs only to select the best/closest TFM and RID.

Reply:
Thanks for the comment! To be follow up in #35283